test/distributed/test_c10d_nccl.py

```diff
@@ -524,75 +552,71 @@ def reduce_scatter(outputs, input_lists, op):
         #   3
 
         # Sum
-        tensor_lists = [
-            [
-                torch.tensor([self.rank * self.num_gpus + i + j]).cuda(i)
-                for j in range(virtual_world_size)
-            ]
-            for i in range(self.num_gpus)
-        ]
+        tensor_lists = []
+        input_per_gpu = []
+
+        for i in range(self.world_size):
+            input_per_gpu.append(torch.tensor([self.rank + i + 1]))
+
+        for gpu in local_device_ids:
+            tensor_lists.append([t.cuda(device=gpu) for t in input_per_gpu])
 
         reduce_scatter(output, tensor_lists, c10d.ReduceOp.SUM)
 
-        for i in range(self.num_gpus):
+        for i in range(num_gpus):
             expected = torch.tensor(
                 [
-                    float(self.num_gpus * (self.num_gpus - 1) / 2)
-                    + (virtual_rank + i) * virtual_world_size
-                ]
-            )
+                    float((1 + self.world_size) * self.world_size / 2)
+                    + self.world_size * self.rank
+                ])
+
+
             # TODO(#38095): Replace assertEqualIgnoreType. See issue #38095
             self.assertEqualIgnoreType(expected, output[i])
 
         # Min
         reduce_scatter(output, tensor_lists, c10d.ReduceOp.MIN)
 
-        for i in range(self.num_gpus):
-            expected = torch.tensor([self.rank * self.world_size + i])
+        for i in range(num_gpus):
+            expected = torch.tensor([self.rank + 1 + i])
             self.assertEqual(expected, output[i])
 
         # Max
         reduce_scatter(output, tensor_lists, c10d.ReduceOp.MAX)
 
-        for i in range(self.num_gpus):
+        for i in range(num_gpus):
             expected = torch.tensor(
-                [self.rank * self.world_size + i + virtual_world_size - 1]
+                [self.rank + self.world_size + i]
             )
             self.assertEqual(expected, output[i])
 
         # Product
-        tensor_lists = [
-            [
-                torch.tensor(
-                    [(self.rank * self.num_gpus + i + j) % virtual_world_size + 1]
-                ).cuda(i)
-                for j in range(virtual_world_size)
-            ]
-            for i in range(self.num_gpus)
-        ]
-
         reduce_scatter(output, tensor_lists, c10d.ReduceOp.PRODUCT)
 
-        for i in range(self.num_gpus):
-            expected = torch.tensor([float(math.factorial(virtual_world_size))])
+        for i in range(num_gpus):
+            prod_val = self.rank + 1
+            for _ in range(self.world_size - 1):
+                prod_val = prod_val * (prod_val + 1)
```

Comment:
This does not seem to look correct. `prod_val` increases not by one but has been multiplied. So if you change the `batch_size` to 4. This test will fail.

Reply:
I didn't see a `batch_size` defined in the test, am i missing something here?

Reply:
Oh, I mean if you change it to 4 GPU, this logic will fail because you are changing the prod_val.

For example if num_gpus = 4 and rank = 0. The expected `prod_val` should be `4! = 24` but this logic will get:
`6 * 7 = 42`.

Reply:
yeah make sense, thanks for pointing out! just updated with the correct formula.